pkg/controller/awsmachine/awsmachine_controller.go

```diff
@@ -69,18 +103,283 @@ type ReconcileAWSMachine struct {
 // Reconcile reads that state of the cluster for a AWSMachine object and makes changes based on the state read
 // and what is in the AWSMachine.Spec
 func (r *ReconcileAWSMachine) Reconcile(request reconcile.Request) (reconcile.Result, error) {
-	// Fetch the AWSMachine instance
-	instance := &infrastructurev1alpha2.AWSMachine{}
-	err := r.Get(context.TODO(), request.NamespacedName, instance)
+	ctx := context.Background()
+
+	// Fetch the AWSMachine instance.
+	awsm := &infrav1.AWSMachine{}
+	err := r.Get(ctx, request.NamespacedName, awsm)
 	if err != nil {
-		if errors.IsNotFound(err) {
-			// Object not found, return.  Created objects are automatically garbage collected.
-			// For additional cleanup logic use finalizers.
+		if apierrors.IsNotFound(err) {
 			return reconcile.Result{}, nil
 		}
-		// Error reading the object - requeue the request.
 		return reconcile.Result{}, err
 	}
 
+	// Store Machine early state to allow patching.
+	patchFrom := client.MergeFrom(awsm.DeepCopy())
+
+	// If the Machine hasn't been deleted and doesn't have a finalizer, add one.
+	if awsm.ObjectMeta.DeletionTimestamp.IsZero() {
+		if !util.Contains(awsm.Finalizers, clusterv1.MachineFinalizer) {
+			awsm.Finalizers = append(awsm.ObjectMeta.Finalizers, clusterv1.MachineFinalizer)
+			if err := r.Client.Patch(ctx, awsm, patchFrom); err != nil {
+				return reconcile.Result{}, errors.Wrapf(err, "failed to add finalizer to AWSMachine %q/%q", awsm.Namespace, awsm.Name)
+			}
+			// Since adding the finalizer updates the object return to avoid later update issues
+			return reconcile.Result{Requeue: true}, nil
+		}
+	}
+
+	// Fetch the Machine.
+	m, err := r.getMachineOwner(ctx, awsm.ObjectMeta)
+	if err != nil {
+		return reconcile.Result{}, err
+	} else if m == nil {
+		klog.Infof("Waiting for Machine Controller to set OwnerRef on AWSMachine %q/%q", awsm.Namespace, awsm.Name)
+		return reconcile.Result{RequeueAfter: 10 * time.Second}, nil
+	}
+
+	// Make sure bootstrap data is available and populated.
+	if m.Spec.Bootstrap.Data == nil || *m.Spec.Bootstrap.Data == "" {
+		klog.Infof("Waiting for bootstrap data to be available on AWSMachine %q/%q", awsm.Namespace, awsm.Name)
+		return reconcile.Result{RequeueAfter: 10 * time.Second}, nil
+	}
+
+	scope, err := actuators.NewMachineScope(actuators.MachineScopeParams{
+		ProviderMachine: awsm,
+		Machine:         m,
+		Client:          r.Client,
+	})
+	if err != nil {
+		return reconcile.Result{}, errors.Errorf("failed to create scope: %+v", err)
+	}
+	defer scope.Close()
+
```

Comment:
Shouldn't this call `r.reconcile()`?

Reply:
Yeah. There is a bunch of missing functionality here and there that should be addressed at some point, I'll add the call